azure_jumpstart_ag/artifacts/PowerShell/Bootstrap.ps1

```diff
@@ -116,115 +179,149 @@ Invoke-WebRequest ($templateBaseUrl + "artifacts/icons/contoso.svg") -OutFile $A
-BITSRequest -Params @{'Uri'='https://wslstorestorage.blob.core.windows.net/wslblob/wsl_update_x64.msi'; 'Filename'="$AgToolsDir\wsl_update_x64.msi"}
-BITSRequest -Params @{'Uri'='https://desktop.docker.com/win/main/amd64/Docker%20Desktop%20Installer.exe'; 'Filename'="$AgToolsDir\DockerDesktopInstaller.exe"}
-BITSRequest -Params @{'Uri'="https://dl.grafana.com/oss/release/grafana-$latestRelease.windows-amd64.msi"; 'Filename'="$AgToolsDir\grafana-$latestRelease.windows-amd64.msi"}
+BITSRequest -Params @{'Uri' = 'https://aka.ms/wslubuntu'; 'Filename' = "$AgToolsDir\Ubuntu.appx" }
+BITSRequest -Params @{'Uri' = 'https://wslstorestorage.blob.core.windows.net/wslblob/wsl_update_x64.msi'; 'Filename' = "$AgToolsDir\wsl_update_x64.msi" }
+BITSRequest -Params @{'Uri' = 'https://desktop.docker.com/win/main/amd64/Docker%20Desktop%20Installer.exe'; 'Filename' = "$AgToolsDir\DockerDesktopInstaller.exe" }
+BITSRequest -Params @{'Uri' = "https://dl.grafana.com/oss/release/grafana-$latestRelease.windows-amd64.msi"; 'Filename' = "$AgToolsDir\grafana-$latestRelease.windows-amd64.msi" }
 
-# Installing tools
-Write-Header "Installing Chocolatey Apps"
-try {
-    choco config get cacheLocation
-}
-catch {
-    Write-Output "Chocolatey not detected, trying to install now"
-    Invoke-Expression ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
-}
+##############################################################
+# Install Chocolatey packages
+##############################################################
+$maxRetries = 3
+$retryDelay = 30  # seconds
 
-Write-Host "Chocolatey Apps Specified"
+$retryCount = 0
+$success = $false
 
-foreach ($app in $AgConfig.ChocolateyAppList) {
-  Write-Host "Installing $app"
-  & choco install $app /y -Force | Write-Output
+while (-not $success -and $retryCount -lt $maxRetries) {
+  try {
+    Write-Header "Installing Chocolatey packages"
+    try {
+      choco config get cacheLocation
+    }
+    catch {
+      Write-Output "Chocolatey not detected, trying to install now"
+      Invoke-Expression ((New-Object System.Net.WebClient).DownloadString($AgConfig.URL.chocoInstallScriptUrl))
+    }
+    
+    Write-Host "Chocolatey packages specified"
+    
+    foreach ($app in $AgConfig.ChocolateyAppList) {
+      Write-Host "Installing $app"
+      & choco install $app /y -Force | Write-Output
+    }
+      
+    # If the command succeeds, set $success to $true to exit the loop
+    $success = $true
+  }
+  catch {
+    # If an exception occurs, increment the retry count
+    $retryCount++
+
+    # If the maximum number of retries is not reached yet, display an error message
+    if ($retryCount -lt $maxRetries) {
+      Write-Host "Attempt $retryCount failed. Retrying in $retryDelay seconds..."
+      Start-Sleep -Seconds $retryDelay
+    }
+    else {
+      Write-Host "All attempts failed. Exiting..."
+      return  # Stop script execution if maximum retries reached      
```

Comment:
What about doing a `throw` here to make sure it's clear that there was an error, and provide some instructions to the user to restart either via running the bootstrap manually, or via delete the RG and start over?

Reply:
also, is there any other cleanup that needs to be done at this point, or is that pointless because all hope is lost?